src/main/java/at/ac/tuwien/kr/alpha/Alpha.java

```diff
@@ -27,201 +27,114 @@
-import org.reflections.scanners.MethodAnnotationsScanner;
-import org.reflections.util.ClasspathHelper;
-import org.reflections.util.ConfigurationBuilder;
-
-import java.io.IOException;
-import java.lang.reflect.Method;
-import java.nio.file.Path;
-import java.util.*;
-import java.util.stream.Stream;
+import at.ac.tuwien.kr.alpha.solver.heuristics.BranchingHeuristicFactory.Heuristic;
 
 public class Alpha {
-	private final Map<String, PredicateInterpretation> predicateMethods = new HashMap<>();
-
-	private final String grounderName;
-	private final String solverName;
-	private final String storeName;
-	private final boolean debug;
-
-	private Solver solver;
-
-	private Program program;
-
-	private long seed;
-
-	public Alpha(String grounderName, String solverName, String storeName, boolean debug) {
-		this.grounderName = grounderName;
-		this.solverName = solverName;
-		this.storeName = storeName;
-		this.debug = debug;
-	}
-
-	public Alpha(String grounderName, String solverName, String storeName) {
-		this(grounderName, solverName, storeName, false);
-	}
 
-	public Alpha(String grounderName, String solverName) {
-		this(grounderName, solverName, "alpharoaming");
-	}
+	private SystemConfig config = new SystemConfig(); // config is initialized with default values
 
-	public Alpha(String grounderName) {
-		this(grounderName, "default");
+	public Alpha(SystemConfig cfg) {
+		this.config = cfg;
 	}
 
 	public Alpha() {
-		this("naive");
-	}
-
-	public long getSeed() {
-		return seed;
-	}
-
-	public void setSeed(long seed) {
-		this.seed = seed;
-	}
-
-	public void setDeterministic() {
-		setSeed(0);
-	}
-
-	public void scan(String base) {
-		Reflections reflections = new Reflections(new ConfigurationBuilder()
-			.setUrls(ClasspathHelper.forPackage(base))
-			.setScanners(new MethodAnnotationsScanner())
-		);
-
-		Set<Method> predicateMethods = reflections.getMethodsAnnotatedWith(Predicate.class);
-
-		for (Method method : predicateMethods) {
-			String name = method.getAnnotation(Predicate.class).name();
-
-			if (name.isEmpty()) {
-				name = method.getName();
-			}
-
-			this.register(method, name);
-		}
 	}
 
-	public void register(Method method, String name) {
-		if (method.getReturnType().equals(boolean.class)) {
-			this.predicateMethods.put(name, new MethodPredicateInterpretation(method));
-			return;
+	public Program readProgram(InputConfig cfg) throws IOException {
+		Program retVal = new Program();
+		if (!cfg.getFiles().isEmpty()) {
+			retVal.accumulate(this.readProgramFiles(cfg.isLiterate(), cfg.getPredicateMethods(), cfg.getFiles()));
 		}
-
-		if (method.getGenericReturnType().getTypeName().startsWith(PredicateInterpretation.EVALUATE_RETURN_TYPE_NAME_PREFIX)) {
-			this.predicateMethods.put(name, new BindingMethodPredicateInterpretation(method));
-			return;
+		if (!cfg.getAspStrings().isEmpty()) {
+			retVal.accumulate(this.readProgramString(StringUtils.join(cfg.getAspStrings(), "\n"), cfg.getPredicateMethods()));
```

Comment:
Not sure if it really makes a difference here, but afaik it is usually better to use `System.lineSeparator()` instead of `"\n"`.

Reply:
changed